xhr/json.any.html WPT test is failing in WebKit

cdumez opened the pull request:
#### c74e2dd4d5886b9a328977a6b46b98a7830b7128
<pre>
xhr/json.any.html WPT test is failing in WebKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=243109">https://bugs.webkit.org/show_bug.cgi?id=243109</a>

Reviewed by Darin Adler.

Force UTF-8 decoding for XHR response when responseType is json.
This aligns our behavior with Blink and Gecko.

* LayoutTests/imported/w3c/web-platform-tests/xhr/json.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/xhr/json.any.worker-expected.txt:
* Source/WebCore/loader/TextResourceDecoder.cpp:
(WebCore::TextResourceDecoder::setEncoding):
(WebCore::TextResourceDecoder::checkForBOM):
* Source/WebCore/loader/TextResourceDecoder.h:
(WebCore::TextResourceDecoder::setAlwaysUseUTF8):
* Source/WebCore/xml/XMLHttpRequest.cpp:
(WebCore::XMLHttpRequest::createDecoder const):

Canonical link: <a href="https://commits.webkit.org/252838@main">https://commits.webkit.org/252838@main</a>
</pre>

darinadler commented on Source/WebCore/loader/TextResourceDecoder.h:
```diff
@@ -68,6 +68,8 @@ class TextResourceDecoder : public RefCounted<TextResourceDecoder> {
     void useLenientXMLDecoding() { m_useLenientXMLDecoding = true; }
     bool sawError() const { return m_sawError; }
 
+    void setAlwaysUseUTF8(bool alwaysUseUTF8) { ASSERT(!strcmp(m_encoding.name(), "UTF-8")); m_alwaysUseUTF8 = alwaysUseUTF8; }
```
I think we should remove the argument here. Only use this when it’s true.

cdumez committed:
xhr/json.any.html WPT test is failing in WebKit
https://bugs.webkit.org/show_bug.cgi?id=243109

Reviewed by Darin Adler.

Force UTF-8 decoding for XHR response when responseType is json.
This aligns our behavior with Blink and Gecko.

* LayoutTests/imported/w3c/web-platform-tests/xhr/json.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/xhr/json.any.worker-expected.txt:
* Source/WebCore/loader/TextResourceDecoder.cpp:
(WebCore::TextResourceDecoder::setEncoding):
(WebCore::TextResourceDecoder::checkForBOM):
* Source/WebCore/loader/TextResourceDecoder.h:
(WebCore::TextResourceDecoder::setAlwaysUseUTF8):
* Source/WebCore/xml/XMLHttpRequest.cpp:
(WebCore::XMLHttpRequest::createDecoder const):

Canonical link: https://commits.webkit.org/252838@main
